### ISSUE TITLE
NVDA processes UIA notification events from all opened edge windows

### ISSUE THREAD

**Issue by beqabeqa473** [nvaccess/nvda#13235]
### Steps to reproduce:
Open several windows of edge browser, for example 20 or more.
Try to download some file from the web.

### Actual behavior:
Your copy of nvda will be blowed up because of too many NEW_DOWNLOAD and DOWNLOAD_COMPLETED UIA notification events.

### Expected behavior:
NVDA should announce starting and completing notification only once per file, only from a current active window.
This doesn't happen if multiple tabs are opened in one window, oly if several windows are opened.

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
alpha-24439,20d5a25d
#### Windows version:
Windows 10 21h2
#### Name and version of other software in use when reproducing the issue:
Microsoft EdgeVersion 97.0.1072.55 (Official build) (64-bit)
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by josephsl** [comment]
Hi, one way to “resolve” (in quotes) this is moving the system focus to an app other than Edge (say, desktop). A more sophisticated solution is checking focus ancestors to make sure NVDA does not announce things from background Edge processes (as in check to make sure the foreground window for the tab where UIA notification is fired is the same as the focused tab). I believe Narrator also suffers from this problem – unlike NVDA, however, Narrator announces UIA notifications from background apps as well. Thanks.

**Comment by beqabeqa473** [comment]
I forgot to say in the initial comment that no, Narrator doesn't experience this problem.

I've tested it with narrator before reporting an issue.

**Comment by josephsl** [comment]
Hi,

Possible cause: focus ancestor check is loose - currently, NVDA only compares app module instances.

Possible solution: tighten focus ancestor checks:

1. Inside the UIA notification event handler, check if the app modules between the UIA object and the focus are the same.
2. If they are the same, make sure the UIA element that sent the notification event is located within the same foreground window as the focused element.

Issues:

1. Scope: the ideal scope is NVDAObjects.UIA.UIA as it can be applied anywhere. Alternaitvely, an pap module based prototype might be easier to test.
2. Tree traversal: what if the notification element does not belong to the same UIA tree as the focused element (ideally, the UIA tree for the foreground window would set foreground window at the root, with other elements branching off it; it might be possible that, due to really bad UIA server implementation, one must traverse all the way up to the desktop object).
3. Performance: it takes time to travel through the UIA tree (especially if an application isn't snappy as before).

The notes above assume that NVDA wants to deal with notifications from the foreground window only.

@LeonardDer and @MichaelDCurran, any thoughts?

Thanks.

**Comment by Adriani90** [comment]
@beqabeqa473 does NVDA crash on your side in this case? I think this issue needs a prioritization somehow.

**Comment by josephsl** [comment]
Hi, cause: limited UIA notification constraints. We can try expanding UIA notification event check to include focus ancestry in addition to app module. Thanks.

**Comment by josephsl** [comment]
Hi,
2024 update: I think an app module based solution would be better as it allows handling background notifications and lets add-ons extend notification checks. The idea would be similar to #10508 except the event to be processed will be UIA notification (a PR to take care of this issue will need to edit UIA handler thread itself).

Thanks.

**Comment by Adriani90** [comment]
@beqabeqa473 is this issue still occuring with NVDA 2024.1 Beta 9?
I gues #14888 has a positive effect on this.

**Comment by seanbudd** [comment]
Could you report this to the Edge developers as well please
